_episodes/l3_03_test_fixture.md

```diff
@@ -0,0 +1,505 @@
+
+```cpp
+TEST_F(Your_test_fixture_class_name, Your_test_name) {
+    // Test logic goes here
+}
+```
+
+Since we now have all the basic tools to create our own test fixtures, let us rewrite the above tests by using a fixture. The code is present in this [2_employeetest.cpp](../code/Chapter3/2_employeetest.cpp). For reference, the tests are shown in the cell below.
+
+```cpp
+// Create a test fixture.
+class EmployeeTestFixture : public::testing::Test {
+    public:
+        Employee employee{"John", 25, 45000, 12, 5000};
+
+};
+
+// Test if we can set the name of an employee.
+TEST_F(EmployeeTestFixture, CanSetName) {
+    employee.setName("John Doe");
+    EXPECT_EQ(employee.getName(), "John Doe");
+}
+
+// Test that the name cannot be empty.
+TEST_F(EmployeeTestFixture, NameCannotBeEmpty) {
+    EXPECT_THROW(employee.setName(""), invalid_argument);
+}
+
+// Test if we can set the age of an employee.
+TEST_F(EmployeeTestFixture, CanSetAge) {
+    employee.setAge(30);
+    EXPECT_EQ(employee.getAge(), 30);
+}
+
+//Test that the tax calculation is correct.
+TEST_F(EmployeeTestFixture, TaxCalculationIsCorrect) {
+    EXPECT_EQ(employee.getTaxAmount(), 7500);
+}
+
+// Check that the net salary is correct.
+TEST_F(EmployeeTestFixture, NetSalaryIsCorrect) {
+    EXPECT_EQ(employee.getNetSalary(), 43500);
+}
+```
+
+## 5. Why do we need setup and teardown in test fixture?
+
+So far, our test fixture class only creates an instance of object for the class under test. In many cases, we often want some common action for all our tests such as adding an entry, connection to a database, response from a site etc. Let us try to understand this with example which will set the background for the `setup` and `teardown` functions.
+
+Let us consider that we are creating a table which will store the details of various employees. The table allows us to add new entries, remove employee from the table, get the number of entries in the table etc. The declaration of table class is given in [employee_table.h](../code/Chapter3/employee_table.h). The definitions of table class is present in [employee_table.cpp](../code/Chapter3/employee_table.cpp). We give the definition of table class below for reference.
+
+```cpp
+void EmployeeTable::addEmployee(const Employee& employee) {
+    employees.push_back(employee);
+}
+
+void EmployeeTable::removeEmployee(const std::string& employeeName) {
+    for (auto it = employees.begin(); it != employees.end(); ++it) {
+        if (it->getName() == employeeName) {
+            employees.erase(it);
+            break;
+        }
+    }
+}
+
+void EmployeeTable::displayEmployeesName() const {
+    cout << "-------------------------------------------------- " << endl;
+    for (const auto& employee : employees) {
+        std::cout << employee.getName() << std::endl;
+    }
+    cout << "-------------------------------------------------- " << endl;
```

Comment:
Same again here. I would just list the function signatures with a one-liner comment explaining what they do, and point the user to the full code.

Reply:
Resolved in df3743fc160148b0b6b2bd59cca4cd511b766008.